### DISCUSSION TITLE
[Bug]: Possible inverter active power issue

### DISCUSSION THREAD

**Discussion by PacmanForever** [wlcrs/huawei_solar#312] [General]
### Describe the issue

Good morning.

First of all, thanks for the integration.

This is my doubt:

When the battery is full, the active power graph of the inverter shows that all solar production is the active power, but while the battery is charging, the graph appears fine.

This is correct? For me the graph should continue as when the battery is charging. I don't understand these values.

![imagen](https://user-images.githubusercontent.com/37216513/222895773-03896d24-19eb-4f2d-a70e-ae0655041b56.png)

Thanks

### Bescribe your Huawei Solar Setup

Inverter Type: SUN2000-5KTL-L1
Inverter Firmware version: V200R001C00SPC125
SDongle present: No
Power meter present: yes
Battery: LUNA2000
Battery Firmware version: V100R002C00SPC116

### How do you connect to the inverter?

Over serial, with a serial-to-WiFi stick

### Upload your Diagnostics File

Drag & Drop your Diagnostics File here.

### Upload your relevant debug logs

```text
I think that it isn't relevant.
```

### Please confirm the following:

- [X] The problem is still present in the latest release of this integration.
- [X] I did not find an existing issue describing this problem.
- [ ] I did upload the diagnostics-file that I could retrieve from  the 'Devices & Services Page'
- [ ] I did upload the relevant debug logs (via 'Enable Debug Logging'-feature or by manually configuring HA logging)

**Comment by PacmanForever** [comment]
Hi again, 

I have added this image of the current state of the entities (the same scenario commented above), notice the circles. 

For me, the two entities should have different values. The first should be current consumption, not production.

![imagen](https://user-images.githubusercontent.com/37216513/222963729-12ebc980-9085-4f7d-af49-92c658fca453.png)

**Reply by JacoboSanchez** [reply]
I am not sure what do you exactly see as incorrect there. Will try to post what I understand (or what I remember from when I configured my integration in the past):

* Input Power is the power produced by your solar panels (DC) 
* Active Power is the power going out of the inverter as AC (Or the DC power going in to the inverter to be transformed to AC¿?)
* In a simple scenario Input Power goes all to the inverter and is transformed in Active Power (except for power transformation losses depending on what is true from previous statement)
* If you have batteries they are in the DC side like the solar panels so when you charge batteries part of the input power goes to battery charge so it does not reach the inverter and it is not summed to the active power
* If you use the batteries the inverter would receive power from batteries so the active power may be higher than input power

This is not an affirmation, just my idea on how things work to validate if that does match or not your scenario. That means that I may be telling everything wrong don't take it as a valid source of information

**Reply by PacmanForever** [reply]
Ok, thank you so much fot this explanation.

I think it would be nice if the contributors and creators of the project were more involved in these discussions.

Only they have exact knowledge of how everything works. Their comments would be of great help and would also improve the adoption of this repository. thank you so much.

**Comment by JacoboSanchez** [comment]
You are talking about a single person that made the effort to have an
integration of it's installation that gives support and features for other
people on topologies he can not even test as he does not own them.

He does much more than expected by giving support on errors of other people
in it's free time to also take care on all the doubts of everyone around
the world.

If we want this integration to succeed we need that users help each other.
The integration's responsibility is just to propagate data provided in
modbus register X named Y by the Huawei documentation  to a sensor called Y
in HA.

What to do with the information provided by your infrastructure in your
home assistant is up you. The one that should give you support on what's
the meaning of one parameter in your Huawei device is Huawei. Of course we
are here to help if we can but the developer of the integration has much
more than enough by providing support and developing features for us in its
free time

**Reply by PacmanForever** [reply]
Yes, I totally agree.

But the internal logic is known only to those who have done it.

I am in no way disparaging his work, just saying that his help would be more clarifying to many, as long as he has free time.

**Comment by jires** [comment]
This is not a bug. 

Names of signal points, like active power, input power etc. has nothing with internal logic of integration to do. They are names of datapoints in Huawei ModBus specification, you should find out in Huawei manuals, what they are represents.  

Short explanation:
- input power: combined power of all dc strings of pv modules connected to the inverter
- charge/discharge power: DC power of charge or drain the battery. Positive value - charge, negative value - drain
- active power: power of AC current generated (or consumed) by inverter. When battery is charging from grid, becomes the negative value. Active power is generated from PV modules (input power) or battery (charge/discharge power) or both.

This should be understand, like the following form:
**[Active power] = [Input power] - [charge/discharge power]**

This formula does not count the inverter losses, therefore values will not be exact. Huawei documentation is unclear, which values are before/after efficiency loss.

**Reply by PacmanForever** [reply]
Thank you very much.
